### ISSUE TITLE
Extend inheritance between view-models

### ISSUE THREAD

**Issue by PatrykSzwer** [Starcounter/Home#294]
I want to create two separate view-models for the same view, and use inheritance for handlers to not duplicate the code. So, Let's say, I have Routes view, that has 2 possible places where It's used. View is the same, most of the handlers will work the same with small exceptions and differences. I would like to be able to create base view-model, with Typed-Json fields which are shared between all derived views/classes and will react on any change made in derived view. Here's small diagram that illustrates what I'm thinking about:
![image](https://user-images.githubusercontent.com/25042256/32164779-cc370b3a-bd60-11e7-82f2-46733d914b18.png)

So It would be awesome if, e.g in `RouteViewForSurface` I would be able to use any derived fields like `Routes [ ]` in the view, and every handler created for `Routes[ ]` change would still be triggered and proceed in base class. It would prevent duplicate handlers code in every derived view. 
If I remember correctly, there was a by design idea, which said that You can't use any property in view, that is not presented in typed-json (even if It's derived). But It's not that bad for this solution, because sometimes You will want to not share every derived property from based typed-json. So you would just have to repeat only properties in typed-jsons but still not the implementation of handling them. WDYT about this functionality? Is it possible to achieve? 

cc @alemoi, @miyconst

**Comment by miyconst** [comment]
@PatrykSzwer thank you for the question/suggestion.

Unfortunately as of now Typed JSON does not support this for various reasons. There are a few possible ways to solve this problem though:

1. Move your logic into a helper class and invoke it from your view-model classes rather than keeping the logic inside view-model.
2. Create a shared view-model and use it within multiple other view-models. [Typed JSON - Supported Datatypes](https://docs.starcounter.io/guides/typed-json/json-by-example#supported-datatypes).

Please close the issue if I answered your question.

P.S. This issue once again reminds us that we have to review our Typed JSON strategy.

**Comment by PatrykSzwer** [comment]
Actually I knew It's right now working like that. My question/idea was more about changing it to be more intuitive and flexible. Do You think We could have functionality like that in the future?

**Comment by miyconst** [comment]
Yes, we definitely can and should have functionality like this, but such request puts entire Typed JSON concept under question mark, and before implementing this or other similar features we should first decide whether we stick with Typed JSON at all.

**Comment by PatrykSzwer** [comment]
Ok, so please have this in mind, that there are people waiting for this functionality 😄.
I'll close this for now. Feel free to re-open when this subject's gonna came up again.

**Comment by chrhol** [comment]
> Unfortunately as of now Typed JSON does not support this for various reasons

Well that is not entirely true. As @PatrykSzwer points out, you have to add the properties you want *exposed to the client* in the inheriting json-by-example. This part was *by-design*. Other inheritance (handlers, using all properties) on the server should work fine. 

So: having a handler for a property defined in the baseclass should work, if the property is exposed in the inherited class (i.e. exposed to client).

I'm not sure exactly what is not working? And reusing one instance (not inheriting) works as well (https://docs.starcounter.io/cookbook/creating-strongly-typed-json-collections, https://docs.starcounter.io/tutorial/expense-tracker#extending-the-view-model)

> Ok, so please have this in mind, that there are people waiting for this functionality 😄.

:+1: 

There are several things that should be reviewed, and several things that could enhance the usage in TypedJSON without a huge amount of work. So we need to take a decision about this soon I think.

**Comment by chrhol** [comment]
Just a small sample about inheritance. 

**BaseJson.json**
```javascript
{
  "Name$": "base"
}
```

**BaseJson.json.cs** 
```csharp
public partial class BaseJson : Json {
    Handle(Input.Name action) {
        action.Value = "Base handler called.";
    }
}
```

**InheritedJson.json**
```javascript
{
  "Name": "Inherited"
}
```

**InheritedJson.json.cs**
```csharp
public partial class InheritedJson : BaseJson {
}
```

Creating an instance of `InheritedJson` and trigger a patch for `Name` will call the handler in the baseclass.

**Comment by miyconst** [comment]
>Creating an instance of `InheritedJson` and trigger a patch for `Name` will call the handler in the baseclass.

But if that method is overridden then it crashes, as far as I remember.

**Comment by PatrykSzwer** [comment]
> But if that method is overridden then it crashes, as far as I remember.

That's what is happening to me right know. Thank You for a tip!
Also thank you @chrhol for an example, I'll ask further questions here when I'll implementing this.

**Comment by PatrykSzwer** [comment]
It would be actually really cool to have it explained in docs as well. WDYT?

**Comment by chrhol** [comment]
> > But if that method is overridden then it crashes, as far as I remember.

> That's what is happening to me right know. Thank You for a tip!

Ahh right, so the problem is overriding a handler in a baseclass then? Not having the handler implemented in the baseclass.

And yes, there seems to be an erronous validation for handlers in the codegenerator that doesn't take in account overriden handlers. I removed the validation (just to test) and the functionality seems to be in place. I can override a handler in the base-class (and of course call it from the inherited handler with `base.Handle(action);`

> It would be actually really cool to have it explained in docs as well. WDYT?

Yes, there's a lot of these hidden things that should be properly documented. But as @miyconst pointed out, let's see if we decide to continue with TypedJSON or not before starting that task.

**Comment by chrhol** [comment]
I couldn't really leave this alone. The fix to allow overriding a handler is actually very very small. Just ignore overriden methods in the codebehind-parser, 3 lines of code (well 4 with the comment). 

```csharp
var isOverriden = node.Modifiers.Any((t) => t.Kind() == SyntaxKind.OverrideKeyword);
if (isOverriden) {
    // We ignore overriden handle methods. They will be dealt with from the baseclass.
    return;
}
```

With this you can mark the handle-method in the baseclass as `virtual` and override it as any other method. Is it ok if I add this fix to `v2.3.1` @miyconst?. This is a relatively safe change. If inheritance is not used, this will not matter or change anything, but might make it at least a little bit less hassle for those that do until we have a decision and a plan forward?

```csharp
partial class BaseJson : Json {
    protected virtual void Handle(Input.Name action) {
         action.Value = "handled in BaseJson";
    }
}

public partial class InheritedJson : BaseJson {
    protected override void Handle(BaseJson.Input.Name action) {
        // Do stuff and then call base Handle
        base.Handle(action);
    }
}
```

**Comment by miyconst** [comment]
@chrhol thank you for the investigation, and feel free to add it to `2.3.1`.
